File: map_reduce_python/counter.py
import re
import sys
import csv
from collections import defaultdict, Counter
from multiprocessing import Queue, Process, Pipe, cpu_count

from utils import timeit


CHUNK_SIZE = 4096 * 2 ** 13  # ~33 MB

M_DATA = 'M_DATA'
M_END = 'M_END'


# message, payload = input

def _count_words(data, counter):
    counter.update(data.split())

def _read_file(name, chunk_size=CHUNK_SIZE):
    with open(name, 'rb') as bin_file:
        while True:
            data = bin_file.read(chunk_size)
            if not data:
                break
            yield data


def count_words(in_queue, to_reducer, to_reader):
    """
    Count words in a string.
    :param in_queue: Input queue
    :param to_reducer: Output queue, send messages to reducer.
    :param to_reader: Output queue, send messages to reducer.
    """
    
    counter = Counter()
    message, payload = in_queue.get()
    while message != M_END:
        _count_words(payload, counter)

        to_reader.put((M_DATA, None))
        message, payload = in_queue.get()

    to_reducer.put((M_DATA, counter))


def reduce(in_queue, pipe_out):
    """
    Reduce the word counts in one dict.
    :param in_queue: Input queue.
    :param pipe_out: Output pipe, send messages to the parent process.
    """
    # counts = defaultdict(int)
    counts = Counter()
    message, payload = in_queue.get()
    while message != M_END:
        counts.update(payload)
        message, payload = in_queue.get()

    pipe_out.send(counts)


def read_file(file_name, nworkers, in_queue, out_queue):
    """
    Read data from a big file
    :param file_name: Name of the input file.
    :param nworkers: Number of workers (word counters).
    :param in_queue: Input queue.
    :param out_queue: Output queue, send messages to word counter.
    """
    file_iterator = _read_file(file_name)

    # First read two full chunks of data
    for i in range(2):
        data = next(file_iterator, None)
        if data is None:
            return # TODO: first M_END should be passed to out_queue

        size = len(data) // nworkers
        for j in range(nworkers):
            out_queue.put((M_DATA, data[j * size: (j + 1) * size]))

    # Then load data and wait...
    for data in _read_file(file_name):
        # Wait for N workers to complete its job before sending more data
        for j in range(nworkers):
            in_queue.get()

        size = len(data) // nworkers
        for j in range(nworkers):
            out_queue.put((M_DATA, data[j * size: (j + 1) * size]))

    for i in range(nworkers):
        out_queue.put((M_END, None))


@timeit
def map_reduce(name):
    nworkers = cpu_count()

    to_counter = Queue()
    to_reader = Queue()
    to_reducer = Queue()
    receiver, sender = Pipe(False)

    reducer = Process(target=reduce, args=(to_reducer, sender), daemon=True)
    reader = Process(target=read_file,
                     args=(name, nworkers, to_reader, to_counter),
                     daemon=True)

    workers = []
    for i in range(nworkers):
        worker = Process(target=count_words,
                         args=(to_counter, to_reducer, to_reader),
                         daemon=True)
        workers.append(worker)

    reader.start()
    reducer.start()
    for w in workers:
        w.start()

    reader.join()

    for w in workers:
        w.join()

    to_reducer.put((M_END, None))
    data = receiver.recv()
    reducer.join()
    return data


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print('Please include the file name as argument.')
        exit(-1)

    data =    map_reduce(sys.argv[1])

    with open('output_py.csv', 'w') as out_file:
        csv_writer = csv.writer(out_file, delimiter=' ')
        for key, val in data.items():
            csv_writer.writerow((key.decode('utf-8'), val))

File: map_reduce_python/cpp/counter.cpp
#include "mtime.hpp"
#include "tbb/concurrent_queue.h"
#include <algorithm>
#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>
#include <fstream>
#include <iostream>
#include <regex>
#include <string>
#include <thread>
#include <unordered_map>
#include <utility>
#include <vector>
#include <deque>

using namespace std;

const size_t CHUNK_SIZE = 4096 * (1 << 13);

const string M_DATA = "M_DATA";
const string M_END = "M_END";

template <class T> using Message = pair<string, T>;

template <class T>
using MessageQueue = tbb::concurrent_bounded_queue<Message<T>>;

using Counter = unordered_map<string, int>;

struct ChunkData {
  string data;
  size_t size;
};

template <typename T> void process_one(MessageQueue<T> &input_queue) {
  while (true) {
    Message<T> message;
    input_queue.pop(message);
    cout << message.second << endl;
  }
}

template <typename T> void process_two(MessageQueue<T> &output_queue) {
  while (true) {
    output_queue.push({"A", "mensaje"});
    this_thread::sleep_for(3s);
  }
}

void read_file(const string &name, const int workers,
               MessageQueue<void *> &in_queue,
               MessageQueue<string> &to_counter) {
  string buffer(CHUNK_SIZE, '\0');
  ifstream file(name);

  for (size_t i = 0; i < 2; i++) {
    if (!file) {
      break;
    }
    file.read(&buffer[0], CHUNK_SIZE);
    const size_t chars_readed = file.gcount();

    const size_t size = (chars_readed + workers) / workers;
    for (size_t j = 0; j < workers; j++) {
      size_t start = j * size;
      size_t end = min((j + 1) * size, chars_readed);
      string data = buffer.substr(start, end);
      to_counter.push({M_DATA, move(data)});
    }
  }

  while (file) {
    file.read(&buffer[0], CHUNK_SIZE);
    const size_t chars_readed = file.gcount();

    for (size_t j = 0; j < workers; j++) {
      Message<void *> message;
      in_queue.pop(message);
    }

    const size_t size = (chars_readed + workers) / workers;
    for (size_t j = 0; j < workers; j++) {
      size_t start = j * size;
      size_t end = min((j + 1) * size, chars_readed);
      string data = buffer.substr(start, end);
      to_counter.push({M_DATA, move(data)});
    }
  }
  for (size_t i = 0; i < workers; i++) {
    to_counter.push({M_END, ""});
  }
  file.close();
}

void _count_words(const string &s, Counter & output) {
  deque<string> strings;
  boost::split(strings, s, boost::is_any_of("\n "));

  for (const auto &str : strings) {
    output[str]++;
  }
}

void count_words(MessageQueue<string> &in_queue,
                 MessageQueue<Counter> &to_reducer,
                 MessageQueue<void *> &to_reader) {
  Counter counter;
  Message<string> message;

  in_queue.pop(message);

  while (message.first != M_END) {
    _count_words(message.second, counter);

    to_reader.push({M_DATA, nullptr});
    in_queue.pop(message);
  }
  to_reducer.push({M_DATA, move(counter)});
}

void reduce(MessageQueue<Counter> &in_queue, MessageQueue<Counter> &out_queue) {
  Counter counter;
  Message<Counter> message;

  in_queue.pop(message);
  while (message.first != M_END) {
    for (const auto &data : message.second) {
      counter[data.first] += data.second;
    }

    in_queue.pop(message);
  }

  out_queue.push({M_DATA, counter});
}

Counter map_reduce(const string &file_name) {
  const size_t nworkers = thread::hardware_concurrency();

  MessageQueue<string> to_counter;
  MessageQueue<void *> to_reader;
  MessageQueue<Counter> to_reducer;
  MessageQueue<Counter> output;

  thread reducer(reduce, ref(to_reducer), ref(output));
  thread reader(read_file, ref(file_name), nworkers, ref(to_reader),
                ref(to_counter));

  vector<thread> workers(nworkers);
  for (auto &worker : workers) {
    worker =
        thread(count_words, ref(to_counter), ref(to_reducer), ref(to_reader));
  }

  reader.join();

  for (auto &worker : workers) {
    worker.join();
  }
  to_reducer.push({M_END, {}});

  Message<Counter> data;
  output.pop(data);

  reducer.join();

  return data.second;
}

int main(int argc, char *argv[]) {
  if (argc < 2) {
    cout << "Please include the file name as argument." << endl;
    return -1;
  }
  Counter data;

  auto time = mtime::mTime([&]() { data = map_reduce(argv[1]); });

  cout << time / 1000.0 << endl;

  ofstream csv_file("cpp_output.csv");

  for (const auto &pair_ : data) {
    csv_file << pair_.first << " " << pair_.second << '\n';
  }
  csv_file.close();
}
